File: saas/static/js/system_health.js
// saas/static/js/system_health.js
import { App } from './state.js';
import { fetchAdminTelemetry, fetchAdminDLQ, retryDLQJob } from './api.js';

let systemHealthPollerInterval = null;

window.showSystemHealthModal = async function() {
    document.getElementById('system-health-modal').style.display = 'flex';
    
    // Clear any existing poller first
    if (systemHealthPollerInterval) {
        clearInterval(systemHealthPollerInterval);
        systemHealthPollerInterval = null;
    }
    
    // Run initial diagnostics immediately
    await pollSystemHealth();
    
    // Set 2-second interval to mirror PySide6 timer
    systemHealthPollerInterval = setInterval(pollSystemHealth, 2000);
};

window.closeSystemHealthModal = function() {
    document.getElementById('system-health-modal').style.display = 'none';
    if (systemHealthPollerInterval) {
        clearInterval(systemHealthPollerInterval);
        systemHealthPollerInterval = null;
    }
};

async function pollSystemHealth() {
    const modal = document.getElementById('system-health-modal');
    if (!modal || modal.style.display === 'none') {
        if (systemHealthPollerInterval) {
            clearInterval(systemHealthPollerInterval);
            systemHealthPollerInterval = null;
        }
        return;
    }

    try {
        const [telemetryRes, dlqRes] = await Promise.all([
            fetchAdminTelemetry(App.token),
            fetchAdminDLQ(App.token)
        ]);

        if (telemetryRes.success && telemetryRes.metrics) {
            const m = telemetryRes.metrics;

            // 1. Update Core Connection Status LEDs (Modern green/red)
            if (m.health) {
                const setLedStyle = (id, healthy) => {
                    const el = document.getElementById(id);
                    if (el) {
                        el.style.backgroundColor = healthy ? '#10b981' : '#ef4444';
                    }
                };

                setLedStyle('led-storage', m.health.storage === 'HEALTHY');
                setLedStyle('led-vector', m.health.vector_db === 'HEALTHY');
                setLedStyle('led-cache', m.health.cache === 'HEALTHY');
                setLedStyle('led-queue', m.health.queue === 'HEALTHY');
                setLedStyle('led-llm', m.health.llm === 'HEALTHY');
            }

            // 2. Circuit Breaker Failover Banner Warning
            const banner = document.getElementById('modal-health-failover-banner');
            if (banner) {
                if (m.circuit_breaker_state === 'OPEN') {
                    banner.classList.remove('hidden');
                } else {
                    banner.classList.add('hidden');
                }
            }

            // 3. Populate Active Worker Threads table
            const workersTbody = document.getElementById('health-active-workers-tbody');
            if (workersTbody) {
                if (!m.active_jobs || m.active_jobs.length === 0) {
                    workersTbody.innerHTML = `
                        <tr>
                            <td colspan="4" style="padding: 15px; text-align: center; color: var(--text-dim);">No active jobs running in worker pool.</td>
                        </tr>
                    `;
                } else {
                    workersTbody.innerHTML = '';
                    m.active_jobs.forEach(job => {
                        const stateColor = job.status === 'RUNNING' ? 'var(--accent-success)' : 'var(--text-dim)';
                        workersTbody.innerHTML += `
                            <tr style="border-bottom: 1px solid rgba(255,255,255,0.05);">
                                <td style="padding: 8px; font-weight: 600;">${job.name}</td>
                                <td style="padding: 8px; font-family: monospace; font-size: 0.8rem; color: var(--accent-cyan);">${job.job_id}</td>
                                <td style="padding: 8px;">${job.task_type}</td>
                                <td style="padding: 8px; color: ${stateColor}; font-weight: 600;">${job.status}</td>
                            </tr>
                        `;
                    });
                }
            }
        }

        // 4. Populate Dead Letter Queue (DLQ) table
        if (dlqRes.success && dlqRes.dlq) {
            const dlqTbody = document.getElementById('health-dlq-tbody');
            if (dlqTbody) {
                if (dlqRes.dlq.length === 0) {
                    dlqTbody.innerHTML = `
                        <tr>
                            <td colspan="5" style="padding: 15px; text-align: center; color: var(--text-dim);">No failed jobs found in DLQ container.</td>
                        </tr>
                    `;
                } else {
                    dlqTbody.innerHTML = '';
                    dlqRes.dlq.forEach(entry => {
                        dlqTbody.innerHTML += `
                            <tr style="border-bottom: 1px solid rgba(255,255,255,0.05);">
                                <td style="padding: 8px; text-align: center;">
                                    <input type="radio" name="modal-dlq-select" value="${entry.job_id}" style="cursor: pointer; accent-color: var(--accent-success);">
                                </td>
                                <td style="padding: 8px; font-family: monospace; font-size: 0.8rem; color: var(--accent-cyan);">${entry.job_id}</td>
                                <td style="padding: 8px;"><span style="background: rgba(231, 76, 60, 0.15); color: #e74c3c; padding: 2px 6px; border-radius: 4px; font-size: 0.8rem; font-weight: 600;">${entry.task_type}</span></td>
                                <td style="padding: 8px; max-width: 300px; white-space: nowrap; overflow: hidden; text-overflow: ellipsis;" title="${entry.error}">
                                    <code style="color: #e74c3c; font-size: 0.8rem;">${entry.error}</code>
                                </td>
                                <td style="padding: 8px; color: var(--text-dim); font-size: 0.8rem;">${entry.timestamp}</td>
                            </tr>
                        `;
                    });
                }
            }
        }
    } catch (e) {
        console.error("[Telemetry UI] Error updating health state:", e);
    }
}

window.retryModalSelectedDLQJob = async function() {
    const selected = document.querySelector('input[name="modal-dlq-select"]:checked');
    if (!selected) {
        alert("Please select a failed job from the Dead Letter Queue table first.");
        return;
    }
    
    const jobId = selected.value;
    const btn = document.getElementById('btn-modal-health-retry');
    const ogHtml = btn.innerHTML;
    
    btn.disabled = true;
    btn.innerHTML = '<i class="fa-solid fa-spinner fa-spin"></i> Dispatched...';
    
    try {
        const res = await retryDLQJob(App.token, jobId);
        if (res.success) {
            alert(res.message || `Job '${jobId}' removed from DLQ and resubmitted successfully.`);
            // Refresh modal instantly
            await pollSystemHealth();
        } else {
            alert("Retry Failed: " + (res.error || "Internal server error"));
        }
    } catch (err) {
        alert("Network error retrying job: " + err.message);
    } finally {
        btn.disabled = false;
        btn.innerHTML = ogHtml;
    }
};

File: desktop/ui/system_health.py
# ui/system_health.py
"""
System Health & Telemetry Standalone Observability Dialog.
Monitors execution thread pool, transparent failover statuses,
and Dead Letter Queue fault states in real time.
"""

from PySide6.QtWidgets import QDialog, QMessageBox, QTableWidgetItem, QHeaderView
from PySide6.QtUiTools import QUiLoader
from PySide6.QtCore import QTimer, Qt
from server.utils.path_utils import get_resource_path
from server.logic.queue.job_queue import JobQueueEngine
from server.logic.services import ServiceRegistry
from desktop.ui.shared_widgets import set_app_icon

class SystemHealthDialog(QDialog):
    """
    Standalone diagnostic popup.
    Features green/red LED connection indicators, background thread execution lists,
    and visual fault-recovery controls for the Dead Letter Queue.
    """
    def __init__(self, parent=None):
        super().__init__(parent)
        set_app_icon(self)
        
        # Load visual layout
        loader = QUiLoader()
        ui_path = get_resource_path("ui_designer/system_health.ui")
        self.ui = loader.load(str(ui_path), self)
        
        if self.ui and self.ui.layout():
            self.setLayout(self.ui.layout())
            self.setFixedSize(self.ui.size())
            
        self.setWindowTitle("📊 System Health & Telemetry Monitor")
        
        # Bind action interactions
        self.ui.btn_close.clicked.connect(self.reject)
        self.ui.btn_retry_task.clicked.connect(self.retry_selected_dlq_job)
        
        # Configure Table Headers layout scaling
        self.ui.table_active_threads.horizontalHeader().setSectionResizeMode(QHeaderView.Stretch)
        self.ui.table_dlq_tasks.horizontalHeader().setSectionResizeMode(QHeaderView.Stretch)
        
        # Hydrate initial layout
        self.ui.lbl_failover_banner.hide()
        self.refresh_diagnostics()
        
        # Connect thread-safe real-time update timer
        self.timer = QTimer(self)
        self.timer.setInterval(2000)  # Pulse updates every 2 seconds
        self.timer.timeout.connect(self.refresh_diagnostics)
        self.timer.start()

    def refresh_diagnostics(self):
        """Updates LEDs and tables with real-time data."""
        # 1. Update Core Connection Status LEDs
        try:
            telemetry = ServiceRegistry.get("telemetry")
            if telemetry:
                health = telemetry.run_health_checks()
                
                def set_led_style(widget, healthy):
                    color = "#10b981" if healthy else "#ef4444"  # Modern green/red
                    widget.setStyleSheet(f"background-color: {color}; border-radius: 8px; min-width: 16px; min-height: 16px; max-width: 16px; max-height: 16px;")
                
                set_led_style(self.ui.lbl_led_storage, health.get("storage") == "HEALTHY")
                set_led_style(self.ui.lbl_led_vector, health.get("vector_db") == "HEALTHY")
                set_led_style(self.ui.lbl_led_cache, health.get("cache") == "HEALTHY")
                set_led_style(self.ui.lbl_led_queue, health.get("queue") == "HEALTHY")
                set_led_style(self.ui.lbl_led_llm, health.get("llm") == "HEALTHY")
                
            # 2. Transparent failover banner warning evaluation
            circuit_breaker = ServiceRegistry.get("circuit_breaker")
            if circuit_breaker:
                if circuit_breaker.state == "OPEN":
                    self.ui.lbl_failover_banner.show()
                else:
                    self.ui.lbl_failover_banner.hide()
        except Exception as e:
            print(f"[Telemetry UI] Error updating health state: {e}")

        # 3. Ingest active worker processing pools
        try:
            queue_engine = JobQueueEngine()
            status = queue_engine.get_queue_status()
            active_list = status.get("processing_jobs", []) + status.get("queued_jobs", [])
            
            self.ui.table_active_threads.setRowCount(len(active_list))
            for row, job in enumerate(active_list):
                self.ui.table_active_threads.setItem(row, 0, QTableWidgetItem(f"WorkerThread-{row+1}"))
                self.ui.table_active_threads.setItem(row, 1, QTableWidgetItem(str(job.get("job_id", ""))))
                self.ui.table_active_threads.setItem(row, 2, QTableWidgetItem(f"Ingest: {job.get('task_type', '')}"))
                self.ui.table_active_threads.setItem(row, 3, QTableWidgetItem(str(job.get("status", "")).upper()))
        except Exception as e:
            print(f"[Telemetry UI] Error updating active threads table: {e}")

        # 4. Ingest Dead Letter Queue table entries
        try:
            queue_engine = JobQueueEngine()
            dlq_entries = queue_engine.get_dlq_entries()
            
            self.ui.table_dlq_tasks.setRowCount(len(dlq_entries))
            for row, entry in enumerate(dlq_entries):
                self.ui.table_dlq_tasks.setItem(row, 0, QTableWidgetItem(str(entry.get("job_id", ""))))
                self.ui.table_dlq_tasks.setItem(row, 1, QTableWidgetItem(str(entry.get("task_type", ""))))
                self.ui.table_dlq_tasks.setItem(row, 2, QTableWidgetItem(str(entry.get("error", ""))))
                self.ui.table_dlq_tasks.setItem(row, 3, QTableWidgetItem(str(entry.get("timestamp", ""))))
        except Exception as e:
            print(f"[Telemetry UI] Error updating DLQ table: {e}")

    def retry_selected_dlq_job(self):
        """Pulls selected job out of Dead Letter Queue and places it back in worker pool."""
        row = self.ui.table_dlq_tasks.currentRow()
        if row < 0:
            QMessageBox.warning(self, "Retry Error", "Please select a failed job from the Dead Letter Queue table first.")
            return
            
        job_id = self.ui.table_dlq_tasks.item(row, 0).text()
        try:
            queue_engine = JobQueueEngine()
            new_job_id = queue_engine.retry_dlq_job(job_id)
            QMessageBox.information(
                self, "Retry Dispatched",
                f"Job '{job_id}' has been removed from DLQ and resubmitted successfully.\n\nNew Background Job ID: {new_job_id}"
            )
            # Rehydrate layout instantly
            self.refresh_diagnostics()
        except Exception as e:
            QMessageBox.critical(self, "Retry Failed", f"An error occurred while retrying the job:\n\n{e}")

    def closeEvent(self, event):
        """Clean up live telemetry polling routine before destruction."""
        self.timer.stop()
        super().closeEvent(event)
